src/engine/ResultTable.cpp

```diff
@@ -1,17 +1,61 @@
-// Copyright 2015, University of Freiburg,
-// Chair of Algorithms and Data Structures.
-// Author: Björn Buchhold (buchhold@informatik.uni-freiburg.de)
+// Copyright 2015 -2022, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors: Björn Buchhold <b.buchhold@gmail.com>
+//          Hannah Bast <bast@cs.uni-freiburg.de>
 
-#include "./ResultTable.h"
+#include "engine/ResultTable.h"
 
 #include <cassert>
 
+#include "global/Id.h"
+#include "global/ValueId.h"
+
+// _____________________________________________________________________________
+Id LocalVocab::getIdAndAddIfNotContained(const std::string& word) {
+  if (constructionHasFinished_) {
+    throw std::runtime_error(
+        "Invalid use of `LocalVocab`: You must not call "
+        "`getIdAndAddIfNotContained` after `endConstructionPhase` has been "
+        "called");
+  }
+  // The following code avoids computing the hash for `word` twice in case we
+  // see it for the first time (note that hashing a string is not cheap). The
+  // return value of the `insert` operation is a pair, where `result.first` is
+  // an iterator to the (already existing or newly inserted) key-value pair, and
+  // `result.second` is a `bool`, which is `true` if and only if the value was
+  // newly inserted.
+  auto result = wordsToIdsMap_.insert(
+      {word,
+       Id::makeFromLocalVocabIndex(LocalVocabIndex::make(words_.size()))});
+  if (result.second) {
+    words_.push_back(word);
+  }
+  return result.first->second;
+}
```

Comment:
In general I have the following question:
Do we need the resolution from Id->Words while the construction is still in place? Otherwise we can save the copies of the strings.
Maybe it would be beneficial to have two types: `LocalVocabCreator` ("construction phase") and then `LocalVocab` (readOnly) for this architecture.
But then again it is not yet clear yet, how "longlived" this design will be (we want to support multiple operations that work on the vocabulary.
The more flexible design would probably to keep the map, but to store only pointers to strings in the map, but that doesn't work, as the HashMap doesn't have pointer stability.
`string_view` also wouldn't work in the vector, because the `std::string` has a small string optimization. So maybe we can keep this for now, and postpone this discussion to the future, when we know, where this implementation is going.

Reply:
But for now,
You could only setup the vector in the `endConstructionPhase()` function, then you can move the strings.

Reply:
There is now a (pointer-stable) hash map from words to IDs and a vector with pointers to the words in the hash map. That way, we do not need to have separate construction phases anymore, at the price of a small overhead in space and time efficiency.